### PR TITLE
alpn: move requested alpn to connection info

### PR DESCRIPTION
**Pull request by botengyao**
Move requested alpn to downstream connection_info for better access from filters.
This is similar as `requestedServerName`

Commit Message:
Additional Description:
Risk Level: low
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: test/common/router/test_corpus/clusterfuzz-testcase-minimized-route_fuzz_test-5085115349270528.fuzz
Comment (yanavlasov) [inline test/common/router/test_corpus/clusterfuzz-testcase-minimized-route_fuzz_test-5085115349270528.fuzz:1]: Was this attached by mistake?

File: envoy/network/socket.h [hunk 1]
```diff
@@ -228,6 +228,11 @@ class ConnectionInfoProvider {
    */
   virtual absl::string_view requestedServerName() const PURE;
 
+  /**
+   * @return requested application protocol(s) (e.g. ALPN in TLS), if any.
+   */
+  virtual const std::vector<std::string>& requestedApplicationProtocols() const PURE;
```
Comment (yanavlasov) [inline envoy/network/socket.h:234]: Can you clarify why you need to add this? Is this to support filter chain matcher?

**Review by yanavlasov — COMMENTED** [review]
/wait-any

File: test/common/router/test_corpus/clusterfuzz-testcase-minimized-route_fuzz_test-5085115349270528.fuzz
Comment (botengyao) [reply test/common/router/test_corpus/clusterfuzz-testcase-minimized-route_fuzz_test-5085115349270528.fuzz:1]: ops, done, thanks!

File: envoy/network/socket.h (hunk 1 above)
Comment (botengyao) [reply envoy/network/socket.h:234]: not really, I think there is no way to let network filter to get the this info due to interface limit to let upstream also consume this info.

**Comment by botengyao** [comment]
/wait 

for a further discussion

**Pull request closed without merging**